components/constellation/constellation.rs

```diff
@@ -1945,6 +1984,170 @@ where
         }
     }
 
+    /// Check the origin of a message against that of the pipeline it came from.
+    /// Note: this is still limited as a security check,
+    /// see https://github.com/servo/servo/issues/11722
+    fn check_origin_against_pipeline(
+        &self,
+        pipeline_id: &PipelineId,
+        origin: &ImmutableOrigin,
+    ) -> Result<(), ()> {
+        let pipeline_origin = match self.pipelines.get(&pipeline_id) {
+            Some(pipeline) => pipeline.load_data.url.origin(),
+            None => {
+                warn!("Received message from closed or unknown pipeline.");
+                return Err(());
+            },
+        };
+        if &pipeline_origin == origin {
+            return Ok(());
+        }
+        Err(())
+    }
+
+    /// Broadcast a message via routers in various event-loops.
+    fn handle_schedule_broadcast(
+        &self,
+        pipeline_id: PipelineId,
+        router_id: BroadcastChannelRouterId,
+        message: BroadcastMsg,
+    ) {
+        if self
+            .check_origin_against_pipeline(&pipeline_id, &message.origin)
+            .is_err()
+        {
+            return warn!(
+                "Attempt to schedule broadcast from an origin not matching the origin of the msg."
+            );
+        }
+        if let Some(routers) = self.broadcast_channel_routers.get(&message.origin) {
+            for (router, (sender, names)) in routers {
+                // Exclude the sender of the broadcast.
+                // Broadcasting locally is done at the point of sending.
+                if router == &router_id {
+                    continue;
+                }
+
+                if names.contains(&message.channel_name) {
```

Comment:
Do you think we'd have to worry about performance here? Should the list of routers be a map of names to routers? I don't really think so, but worth flagging.

Reply:
I don't know really. You mean the call to `contains` would be negative most of the time so worth skipping? 

We'd have to either clone the sender for each name associated with a router, or make it a map of name to router ids, and then have another map from router id to sender.

I think this is the most "logical" way to express the logic, we'd have to have a reason to change it for performance reasons...

Reply:
I think you're right actually, it might be better to have one data structure represent the "active channels" per origin/name and the relevant routers, and then have another one represent the router -> channel to actual do the broadcast. Now it's all nested which makes it maybe more complicated. Should address any potential performance issues too...